### DISCUSSION TITLE
How do I insert a table in a two-column PDF layout?

### DISCUSSION THREAD

**Discussion by cbrnr** [quarto-dev/quarto-cli#11669] [Q&A] [answered]
I would like to insert a computed table into a two-column PDF layout. Consider the following example:

````qmd
---
title: "Two-Column Layout with Table"
format:
  pdf:
    documentclass: article
    classoption: twocolumn
---

```{r setup, include=FALSE}
library(knitr)
```

This is some introductory text.

```{r}
kable(head(cars))
```
````

This produces the following error:

```
ERROR: 
compilation failed- error
Package longtable Error: longtable not in 1-column mode.

See the longtable package documentation for explanation.
Type  H <return>  for immediate help.
 ...                                              
                                                  
l.207 \begin{longtable}
                       []{@{}rr@{}}
```

It seems like the `longtable` LaTeX package is not compatible with multi-column layouts according to [this answer](https://tex.stackexchange.com/a/161455), which suggests to use the `xtab` package instead. Is this something I can do in Quarto, or is this only possible in `knitr::kable()` (or is this perhaps a bug that should be reported)?

I'm using Quarto 1.6.39 on Arch Linux (but the problem also occurs on other OSs like Windows).

**Comment by mcanouil** [comment]
Quarto does not control what third party packages/libraries emit.
Quarto will only try to detect when `longtable` is used to replace it with `longtable*` to make it work in Quarto cross-reference system.

For LaTeX packages conflicts, you need to handle them independently of Quarto.
Look at `knitr::kable()` documentation, for instance `knitr::kable(..., format = "latex", table.envir = "table*")` or `knitr::kable(..., format = "latex", longtable = FALSE)`, etc.
This is really not up to Quarto.


Sidenote: prefer YAML syntax in Quarto

````
```{r setup, include=FALSE}
````

to

````
```{r}
#| label: setup
#| include: false
````

**Reply by cbrnr** [reply]
But I don't understand why `longtable` is used in the first place. The default of `knitr::kable()` is `tabular`:

```
> knitr::kable(head(cars), format="latex")

\begin{tabular}{r|r}
\hline
speed & dist\\
\hline
4 & 2\\
\hline
4 & 10\\
\hline
7 & 4\\
\hline
7 & 22\\
\hline
8 & 16\\
\hline
9 & 10\\
\hline
\end{tabular}
```

**Reply by mcanouil** [reply]
I can't speak for `knitr` internals in regards to Quarto.
@cderv is the expert on that part.

**Reply by cscheid** [reply]
@cbrnr You can use the `keep-tex: true` and `keep-md:  true` debugging aids:

The output of `keep-md: true` offers a hint here:

````
---
title: "Two-Column Layout with Table"
keep-md: true
keep-tex: true
format:
  pdf:
    documentclass: article
    classoption: twocolumn
---



This is some introductory text.

::: {.cell}

```{.r .cell-code}
kable(head(cars))
```

::: {.cell-output-display}


| speed| dist|
|-----:|----:|
|     4|    2|
|     4|   10|
|     7|    4|
|     7|   22|
|     8|   16|
|     9|   10|


:::
:::
````

You can see that knitr is emitting Markdown tables, which Pandoc by default renders using longtable. So the issue is that Pandoc's default table rendering code isn't compatible with `twocolumn`.

**Reply by cbrnr** [reply]
Thanks @cscheid, I guess I'll ask the Pandoc devs if there's anything that can be done.

But don't you think this is also a problem for Quarto? I can't be the first person trying to use Quarto to create a two-column PDF with a table?

**Reply by cbrnr** [reply]
Related issue (created Oct 2013 👀): https://github.com/jgm/pandoc/issues/1023

**Reply by cscheid** [reply]
> But don't you think this is also a problem for Quarto?

Yes, it is. But the solution I would like to see implemented is for the markdown-to-latex community to move to https://ctan.math.washington.edu/tex-archive/macros/latex/contrib/tabularray/tabularray.pdf

Quarto already does a large number of ugly workarounds for `longtable` because of its quirks (I know they're ugly because I wrote them :)), and I can no longer justify the decrease in overall code quality to add more such workarounds.

**Reply by cbrnr** [reply]
😆❤️ Do you see this transition happening anytime soon? Or should I try Typst instead (maybe it just works)?

**Reply by cscheid** [reply, marked as answer]
I think that if you ask knitr for latex output, then it emits tabular, which _does_ work.

What you need to avoid in the emitted LaTeX code is:

- `longtable` (because it plays badly with LaTeX crossreferences _and_ two-column mode)
- `table` (because it plays badly with LaTeX crossreferences)

**Reply by cbrnr** [reply]
Thanks @cscheid! Using `knitr::kable(..., format="latex")` indeed works! 🚀

**Reply by garvanwalshe** [reply]
Hi, this is great, is there an equivalent for people who aren't using R? I'm using python and jupyter.

**Reply by cderv** [reply]
@garvanwalshe it all depends on how you create your table in Python, and which format you target. 

If you want to target PDF with LaTeX like in here, then you need to output a LaTeX table that avoids `longtable` and `table` as explained. 

For example, Pandas has an output format to be LaTeX. 
https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_latex.html

Example:

````markdown
---
title: "Reproducible Quarto Document"
format: pdf
keep-md: true
---

```{python}
#| output: asis
import pandas as pd
df = pd.DataFrame(dict(name=['Raphael', 'Donatello'],
                       age=[26, 45],
                       height=[181.23, 177.65]))
print("```{=latex}")
print(df.to_latex(index=False,
                  formatters={"name": str.upper},
                  float_format="{:.1f}".format,
))
print("```")
```
````

This is using `tabular` for LaTeX tables, so it could works. I think polars using `great_tables` otherwise, but outputing `longtable` or `table` environment, which is not great in quarto context. 

Hope it helps

**Comment by zinc75** [comment]
13 years later, pandoc seems to have a solution for this longtable problem in twocolumns mode : https://github.com/jgm/pandoc/commit/3ecd95bf00a595e208df2b0d071e0e296d9d07b3 https://github.com/jgm/pandoc/issues/1023#issuecomment-4936486618 . However, this seems to be active only with pandoc 3.10.1 (version 3.10 included in quarto 1.10.18), and needs the use of a small lua filter that quarto could handle.

**Reply by cderv** [reply]
I have opened an issue to see how we could leverage that once we update to 3.10.1+

- https://github.com/quarto-dev/quarto-cli/issues/14846

Not as easy as it seems unfortunately due to our cross reference table work.
